iac-common/src/main/java/org/sonar/iac/common/yaml/tree/YamlTreeMetadata.java

```diff
@@ -29,69 +33,134 @@
-    if (commentLines == null) {
-      return Collections.emptyList();
+  public static class Builder {
+
+    private Node startNode;
+    private Node endNode;
+    private String tag;
+    private List<Comment> comments;
+
+    public Builder fromNode(Node node) {
+      this.startNode = node;
+      this.endNode = node;
+      return this;
     }
-    List<Comment> comments = new ArrayList<>();
-    for (CommentLine comment : commentLines) {
-      comments.add(comment(comment));
+
+    public Builder fromNodes(Node startNode, Node endNode) {
+      this.startNode = startNode;
+      this.endNode = endNode;
+      return this;
     }
-    return comments;
-  }
 
-  private static Comment comment(CommentLine comment) {
-    // We prefix the comment value with # as it is already stripped away when arrive at this point.
-    return new CommentImpl('#' + comment.getValue(), comment.getValue(), range(comment.getStartMark(), comment.getEndMark()));
-  }
+    public Builder withTag(String tag) {
+      this.tag = tag;
+      return this;
+    }
 
-  public static TextRange range(Node node) {
-    return range(node.getStartMark(), node.getEndMark());
-  }
+    public Builder withComments(List<CommentLine> commentLines) {
+      this.comments = comments(commentLines);
+      return this;
+    }
+
+    public YamlTreeMetadata build() {
+      if (tag == null) {
+        tag = tag(startNode);
+      }
+      if (comments == null) {
+        comments = comments(startNode);
+      }
 
-  private static TextRange range(Optional<Mark> startMark, Optional<Mark> endMark) {
-    if (startMark.isEmpty()) {
-      throw new ParseException("Nodes are expected to have a start mark during conversion", null, null);
+      var range = TextRanges.merge(range(startNode), range(endNode));
+      var startPointer = pointer(startNode.getStartMark().orElse(null));
+      var endPointer = startPointer;
+      if (endNode.getEndMark().isPresent()) {
+        endPointer = pointer(endNode.getEndMark().orElse(null));
+      }
+      return new YamlTreeMetadata(tag, range, startPointer, endPointer, comments);
     }
 
-    int startLine = startMark.get().getLine() + 1;
-    int startColumn = startMark.get().getColumn();
+    public static String tag(Node node) {
+      return node.getTag().getValue();
+    }
 
-    // endMark is not present. This happens for example when we have a file with only a comment.
-    // in that case, the root node will be an empty MappingNode with only a startMark to which the comment is attached
-    return endMark.map(mark -> TextRanges.range(startLine, startColumn, mark.getLine() + 1, mark.getColumn()))
-            .orElseGet(() -> TextRanges.range(startLine, startColumn, startLine, startColumn));
-  }
+    private static int pointer(@Nullable Mark mark) {
+      if (mark == null) {
+        throw new ParseException("Nodes are expected to have a start mark during conversion", null, null);
+      }
```

Comment:
To cover this, I would have to write an artificial test case. There is no case that this will be null. But it's the result from SnakeYaml. So better safe than sorry.